Home Work №3

kitzsable opened the pull request.

Commit:
Home Work №3

pavelsilent commented on unit4/src/main/java/pro/sisit/adapter/impl/CSVAdapter.java:
```diff
@@ -18,15 +29,124 @@ public CSVAdapter(Class<T> entityType, BufferedReader reader,
+    private int countOfLen() throws IOException {
+        int count = 0;
+        String s;
+        reader.mark(100000);
```
100000 - можно убрать в константу.

pavelsilent commented on unit4/src/main/java/pro/sisit/CSVDelimiter.java:
```diff
@@ -0,0 +1,8 @@
+/**
+ * Интерфейс, содержащий разделитель в CSV файлах
+ */
+public interface CSVDelimiter {
```
Немного странно для разделителя использовать интерфейс и тем более имплементить от него что-то. Интерфейс - это "контракт", например, он должен наделять объект чем-то. По Вашей логике (и названию) - что класс, реализующий является разделителем :-) 

Что можно сделать:
 - создать обычный класс с static полем
 - создать класс с настройками для парсинга csv и подавать его адаптеру. Тогда мы бы могли подавать разные настройки в адаптер.

Но то, что Вы подумали о параметризации разделителя - уже очень хорошо.

pavelsilent commented on unit4/src/main/java/pro/sisit/adapter/impl/CSVAdapter.java:
```diff
@@ -2,11 +2,22 @@
 
-public class CSVAdapter<T> implements IOAdapter<T> {
+// Класс, который будет работать с адаптером,
+// должен реализовывать интерфейс FieldPlaceholder
```
хм, FieldPlaceholder. Я не вижу у вас в коде такой интерфейс. Можно подробнее о чем речь ?

kitzsable replied:
Забыл исправить комментарий, имелся ввиду CSVObject

pavelsilent replied:
Принято. Я закончил смотреть работу - сейчас напишу итог.

pavelsilent commented on unit4/src/main/java/pro/sisit/adapter/impl/CSVAdapter.java:
```diff
@@ -18,15 +29,124 @@ public CSVAdapter(Class<T> entityType, BufferedReader reader,
+            // Считывание строки
+            String string = reader.readLine();
+            // Если ничего не считалось, то выбрасывается исключение, т.к. строки нет
+            if (string==null) throw new RuntimeException(String.format("Файл не имеет строки под номером %d", index));
```
Даже однострочный if - нужно обрамлять фигурными скобками. Нам самим же проще будет ориентироваться в коде:

```
if (string==null) {
   throw new RuntimeException(
      String.format("Файл не имеет строки под номером %d", index));
}

```

pavelsilent commented on unit4/src/main/java/pro/sisit/adapter/impl/CSVAdapter.java:
```diff
@@ -18,15 +29,124 @@ public CSVAdapter(Class<T> entityType, BufferedReader reader,
     public T read(int index) {
-        throw new RuntimeException("Метод read не реализован");
+        // Создание переменной типа T
+        T newT = null;
```
newT - так себе название для переменной. Можно понять о чем речь, но лучше назвать более адекватно.

pavelsilent commented on unit4/src/main/java/pro/sisit/adapter/impl/CSVAdapter.java:
```diff
@@ -18,15 +29,124 @@ public CSVAdapter(Class<T> entityType, BufferedReader reader,
+            // Запоминаем начальную позицию
+            reader.mark(100000);
+            // Получение объекта типа T, с пмощью пустого конструктора
+            newT = entityType.getDeclaredConstructor().newInstance();
```
Рефлексию не рекомендуется использовать без необходимости. Исправлять не нужно, но подумайте как можно без нее обойтись. Более подробно напишу в итоговом комменте.

pavelsilent commented on unit4/src/main/java/pro/sisit/adapter/impl/CSVAdapter.java:
```diff
@@ -18,15 +29,124 @@ public CSVAdapter(Class<T> entityType, BufferedReader reader,
+     */
+    private List<String> getLines() throws IOException {
+        List<String> listOfLines = new ArrayList<>();
+        reader.mark(100000);
```
в константу

pavelsilent commented on unit4/src/main/java/pro/sisit/adapter/impl/CSVAdapter.java:
```diff
@@ -18,15 +29,124 @@ public CSVAdapter(Class<T> entityType, BufferedReader reader,
+     * чтобы дальнейшая запись шла в конец файла
+     * @throws IOException Ошибка работы с файлом
+     */
+    private void reWrite() throws IOException {
```
Немного странный метод, и, немного странно вызывать его прям в конструкторе. Мы считали все строки и, заново записали их в файл, так? Для чего?
 - Представьте что у нас файл на 1000000 (или еще больше) строк. Как быстро пройдет инициализация адаптера?
 - Вы говорите - это "для того, чтобы дальнейшая запись шла в конец файла". А если мне нужно писать не в конец? Правильно, всеравно придется перематывать! 

Подумайте еще раз, действительно ли существование и вызов этого метода - необходимо?

pavelsilent commented on unit4/src/main/java/pro/sisit/adapter/impl/CSVAdapter.java:
```diff
@@ -18,15 +29,124 @@ public CSVAdapter(Class<T> entityType, BufferedReader reader,
     @Override
-    public int append(T entity) {
-        throw new RuntimeException("Метод append не реализован");
+    public int append(CSVObject entity) {
```
Неплохо бы добавить проверку entity на null.

pavelsilent commented on unit4/src/main/java/pro/sisit/adapter/impl/CSVAdapterLog.java:
```diff
@@ -0,0 +1,25 @@
+import pro.sisit.adapter.IOAdapter;
+import pro.sisit.model.CSVObject;
+
+public class CSVAdapterLog<T extends CSVObject> implements IOAdapter<T> {
```
К адаптеру замечаний - нет. Все отлично.

pavelsilent commented on unit4/src/main/java/pro/sisit/model/Author.java:
```diff
@@ -1,17 +1,44 @@
+     */
+    @Override
+    public void fillField(List<String> list) {
+        if (list.size() == this.getClass().getDeclaredFields().length) {
```
Опять - рефлексия! :-)

pavelsilent commented on unit4/src/main/java/pro/sisit/model/Book.java:
```diff
@@ -1,21 +1,50 @@
+     */
+    @Override
+    public String getCSVString() {
+        return String.format("%s%s%s%s%s%s%s", name, DELIMITER, author, DELIMITER, genre, DELIMITER, isbn);
```
Выглядит - монструозно. А если в объекте будет 100 полей - как будет выглядеть template?

pavelsilent commented on unit4/src/test/java/pro/sisit/adapter/CSVAdapterTest.java:
```diff
@@ -2,91 +2,186 @@
+        File librariesFile = Paths.get("test-library-file.csv").toFile();
+        librariesFile.delete();
+        authorsFile.delete();
+        booksFile.delete();
```
Было бы неплохо проверить, что файлы удалились.

pavelsilent commented on unit4/src/test/java/pro/sisit/adapter/CSVAdapterTest.java:
```diff
@@ -2,91 +2,186 @@
 public class CSVAdapterTest {
 
     @Before
     public void createFile() {
```
Это замечание не критичное, его можно не править. Но, немного странно готовить данные для проверки тем сервисом, который мы и собираемся проверять!

В методе createFile предполагалось что Вы будете либо создавать новый файл с данными. (причем не используя csvAdapter, т.к. мы его проверять будем), а в методе delete - удалять его. Есть два варианта:
 - Первый вариант. У вас в папке с ресурсами для тестов уже есть подготовленный файл с записями. В этом методе вы его содержимое копируете во временный файл, и далее по тесты работаете с временным файлом. Затем, после того как тесты прошли - в методе deleteFile - вы удаляете временный файл.
 - Второй вариант. В методе createFile вы создаете и заполняете файл для тестов, но, повторюсь - не используя csvAdapter. Например, можете просто открыть поток и записать сразу готовые строки в файл (сразу с разделителями). В методе deleteFile - вы удаляете этот файл.

pavelsilent commented on unit4/src/test/java/pro/sisit/adapter/CSVAdapterTest.java:
```diff
@@ -2,91 +2,186 @@
+             BufferedWriter libraryWriter = new BufferedWriter(new FileWriter(librariesFile, true))) {
+
+            // Тесты сущности Author
+            IOAdapter<Author> authorAdapter = new CSVAdapterLog<>(new CSVAdapter<>(Author.class, authorReader, authorWriter));
```
Выглядит немного громоздко, можно разделить на строки.

pavelsilent commented:
Очень хорошая работа. Из критичного, необходимого для исправления, только одно:
 - reWrite() - непонятно зачем, подумайте что можно с ним сделать. Более подробно - в комменте к коду.

Улучшения:
 - модели, так или иначе, знают о csv, знают о разделителе, знают о том кто и как будет с ними работать. Хотелось бы избавить модель от этих знаний. Подумайте как можно это сделать. Если ничего не придумаете - можно написать в личку в телеграм (Pavel Brychev), я дам доп. подсказку.

 - много рефлексии. Не рекомендуется использовать рефлексию. Погуглите про паттерн проектирования "Фабрика", подумайте как можно было бы применить ее здесь.

- замечание к тестам о том, что они готовят себе файлы для проверки, тем же механизмом, который собираются проверять. Более подробно - в комменте к коду.

Исправление улучшений, а так же мелких помарок к коду - на Ваше усмотрение. Если что-то будет непонятно - пишите.

pavelsilent commented:
Напоминаю, что по прежнему есть одно критичное замечание, которое надо исправить. Тогда работа будет принята.
Также есть ряд улучшений - фиксы на Ваше усмотрение.

Commit:
Рефлексия заменена Фабрикой

Метод reWrite() удалён, т.к. запись в конец происходит автоматически при передачи в конструктор FileWriter вторым параметром значения true

Разделитель перенесен в интерфейс CSVObject

Изменен способ сооздания файлов перед тестированием

Добавлена проверка удаления файлов

Значение readAheadLimit вынесено в константу

Исправлено форматирование конструкции if

pavelsilent commented on unit4/src/main/java/pro/sisit/adapter/impl/CSVAdapter.java:
```diff
@@ -45,45 +42,44 @@ public CSVAdapter(Class<T> entityType, BufferedReader reader,
-            newT = entityType.getDeclaredConstructor().newInstance();
+            reader.mark(readAheadLimit);
+            // Получение объекта
+            switch (entityType.getSimpleName()) {
```
По идее должно хватить одного switch - именно в фабрике. Здесть просто подавать строковое значение, считанное из файла - в фабрику и получать из нее объект.
Т.е. формально значения енама могут совпадать с значениями, которые будут получены из файла. Ну либо можно добавить доп. поле в енам.

kitzsable replied:
Поправил, убрал дубликат

pavelsilent commented:
26.03.20 - посмотрел работу. Все круто! Единственное замечание к дублю switch. Попробуйте оставить его **только** в фабрике. А так - все идеально.

Commit:
Удалена дубликация switch

pavelsilent commented on unit4/src/main/java/pro/sisit/model/CSVObjectType.java:
```diff
@@ -1,7 +0,0 @@
-package pro.sisit.model;
```
Да можно было не удалять класс. Енамы можно расширять. Например так:

```
public enum CSVObjectType {
    AUTHOR("AUTHOR", "Author"),
    BOOK("BOOK", "Book"),
    LIBRARY("LIBRARY", "Library");

    @Getter
    private final String code;
    @Getter
    private final String name;

    CSVObjectType(String code, String name) {
        this.code = code;
        this.name = name;
    }
}
```

и соответсвенно в коде использовать:
`CSVObjectType.AUTHOR.getName()`

pavelsilent commented:
Все хорошо. Написал замечание к енаму. Можно не править, просто разберитесь как это работает.
Можно сливать мердж.
